exchanges/stream/buffer/buffer.go

```diff
@@ -42,246 +77,345 @@ func (w *Orderbook) Update(u *Update) error {
+		insertUpdatesAsk(updts.Asks, &o.ob.Asks)
+	case UpdateInsert:
 	updateBids:
-		for x := range u.Bids {
-			for y := range o.Bids {
-				if o.Bids[y].ID == u.Bids[x].ID {
-					o.Bids[y].Amount = u.Bids[x].Amount
+		for x := range updts.Bids {
+			for target := range o.ob.Bids { // First iteration finds ID matches
+				if o.ob.Bids[target].ID == updts.Bids[x].ID {
+					if o.ob.Bids[target].Price != updts.Bids[x].Price {
+						// Price change occurred so correct bid alignment is
+						// needed - delete instance and insert into correct
+						// price level
+						o.ob.Bids = append(o.ob.Bids[:target], o.ob.Bids[target+1:]...)
+						break
+					}
+					o.ob.Bids[target].Amount = updts.Bids[x].Amount
 					continue updateBids
 				}
 			}
-			o.Bids = append(o.Bids, u.Bids[x])
+			insertBid(updts.Bids[x], &o.ob.Bids)
 		}
-
 	updateAsks:
-		for x := range u.Asks {
-			for y := range o.Asks {
-				if o.Asks[y].ID == u.Asks[x].ID {
-					o.Asks[y].Amount = u.Asks[x].Amount
+		for x := range updts.Asks {
+			for target := range o.ob.Asks {
+				if o.ob.Asks[target].ID == updts.Asks[x].ID {
+					if o.ob.Asks[target].Price != updts.Asks[x].Price {
+						// Price change occurred so correct ask alignment is
+						// needed - delete instance and insert into correct
+						// price level
+						o.ob.Asks = append(o.ob.Asks[:target], o.ob.Asks[target+1:]...)
+						break
+					}
+					o.ob.Asks[target].Amount = updts.Asks[x].Amount
 					continue updateAsks
 				}
 			}
-			o.Asks = append(o.Asks, u.Asks[x])
+			insertAsk(updts.Asks[x], &o.ob.Asks)
+		}
+	default:
+		return fmt.Errorf("invalid action [%s]", updts.Action)
+	}
+	return nil
+}
+
+// applyUpdates amends amount by ID and returns an error if not found
+func applyUpdates(updts, book []orderbook.Item) error {
+updates:
+	for x := range updts {
+		for y := range book {
+			if book[y].ID == updts[x].ID {
+				book[y].Amount = updts[x].Amount
+				continue updates
+			}
 		}
+		return fmt.Errorf("update cannot be applied id: %d not found",
+			updts[x].ID)
 	}
+	return nil
 }
 
-// LoadSnapshot loads initial snapshot of ob data, overwrite allows full
-// ob to be completely rewritten because the exchange is a doing a full
-// update not an incremental one
-func (w *Orderbook) LoadSnapshot(newOrderbook *orderbook.Base) error {
-	if len(newOrderbook.Asks) == 0 || len(newOrderbook.Bids) == 0 {
-		return fmt.Errorf("%v snapshot ask and bids are nil", w.exchangeName)
+// deleteUpdates removes updates from orderbook and returns an error if not
+// found
+func deleteUpdates(updt []orderbook.Item, book *[]orderbook.Item, bypassErr bool) error {
+updates:
+	for x := range updt {
+		for y := range *book {
+			if (*book)[y].ID == updt[x].ID {
+				*book = append((*book)[:y], (*book)[y+1:]...) // nolint:gocritic
+				continue updates
+			}
+		}
+		// bypassErr is for expected duplication from endpoint.
+		if !bypassErr {
+			return fmt.Errorf("update cannot be deleted id: %d not found",
```

Comment:
I modified this to include pair and asset:
```
routines.go exchange Bitfinex websocket error - bitfinex_websocket.go updating orderbook error: marginfunding USD update cannot be 1052932921 %!v(MISSING) deleted id: %!d(MISSING) not found
```
Looks like margin funding bitfinex is having issues